# Projeto Final de Machine Learning
Feito por: _Henrique Bucci_ e _Marcelo Alonso_

Dados e informações: https://www.kaggle.com/datasets/marcopale/housing/data

**Perguntas**  
- Posso remover o PID?
    - **R:** Sim
- Posso criar colunas a partir de contas de outras antes de fazer a seleção?
    - **R:** Sim
- Se eu aplicar PolynomialFeatures nos dados, eles também contam como features para a contagem?
    - **R:** Fazer PolyFeatures depois de selecionar as features
- Posso utilizar correlação na análise exploratória?
    - **R:** Pode, mas é "inútil"
- Posso utilizar métodos de clustering na pipeline para incluir a classificação como uma nova feature?
    - **R:** SoftMax no resultado do Kmeans para exagerar a classe mais próxima.
- Posso utilizar algum método de Dimensionality Reduction (ex: PCA) para me ajudar a escolher as features?
    - **R:** Sim.


Testar stacking: Treinar diversos modelos e treinar um modelo final com os predicts destes modelos.

#### ANOTAÇÕES
Utilizar LASSO para seleção de features.

Regressão linear para ignorar outliers.

RANSAC -> regressao linear que ignora outliers

## Etapa 0

Nesta etapa, iremos:
- Importar bibliotecas
- Carregar os dados
- Verificar se existem colunas que não fazem sentido serem colocadas no dataset final (como ID ou algum outro tipo de identificador arbitrário), olhando apenas a descrição das colunas.
- Separar o dataset em Treino-Teste

### Bibliotecas

In [2]:
import pandas as pd
from datetime import datetime
from utils import *


from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.dummy import DummyRegressor
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, FunctionTransformer
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import LinearSVR, SVR

### Constantes

In [3]:
SEED = 420

### Carregamento e Pré-processamento dos Dados

In [4]:
dataset = load_data()
dataset.head()

Order        PID  MS.SubClass MS.Zoning  Lot.Frontage  Lot.Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot.Shape Land.Contour  ... Pool.Area Pool.QC  Fence Misc.Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc.Val Mo.Sold Yr.Sold Sale.Type  Sale.Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [5]:
"""
Neste caso, a presença de duplicatas não seria intencional, uma vez que cada casa deveria ser única.
Portanto, vamos removê-las.
"""
print(f"Total de linhas antes de remover duplicatas: {dataset.shape[0]}")
dataset.drop_duplicates(inplace=True) # Removendo duplicatas
print(f"Total de linhas depois de remover duplicatas: {dataset.shape[0]}")

Total de linhas antes de remover duplicatas: 2930
Total de linhas depois de remover duplicatas: 2930


In [6]:
"""
Como a primeira coluna é o ID da observação e a segunda é um identificador, podemos removê-las, uma vez que estes são valores arbitrários.
"""
dataset = dataset.iloc[:, 2:] # Estamos removendo as duas primeiras colunas, que são o ID e o PID (Parcel identification number)

### Criando novas features

Ao analisarmos as features da forma descrita acima, vimos espaço para a criação de novas features que podem ser úteis na modelagem dos dados:
- **Tot Lot Area** : `Lot Frontage + Lot Area`
- **Bsmt Tot Bath** : `Bsmt Full Bath + 0.5*Bsmt Half Bath`
- **Garage Area/Car** : `Garage Area / Garage Cars`
- **Tot Porch SF** : `Open Porch SF + Enclosed Porch + 3Ssn Porch + Screen Porch`
- **Date Sold** : `timestamp(Month Sold, Year Sold)`

In [7]:
dataset.loc[:, 'Tot.Lot.Area'] = dataset.loc[:, 'Lot.Frontage'] + dataset.loc[:, 'Lot.Area']
dataset.loc[:, 'Bsmt.Tot.Bath'] = dataset.loc[:, 'Bsmt.Full.Bath'] + 0.5*dataset.loc[:, 'Bsmt.Half.Bath']
dataset.loc[:, 'Garage.Area/Cars'] = dataset.loc[:, 'Garage.Area'] / dataset.loc[:, 'Garage.Cars']
dataset.loc[:, 'Tot.Porch.SF'] = dataset.loc[:, 'Open.Porch.SF'] + dataset.loc[:, 'X3Ssn.Porch'] + dataset.loc[:, 'Enclosed.Porch'] + dataset.loc[:, 'Screen.Porch']
dataset.loc[:, 'Date.Sold'] = pd.to_datetime(dict(year=dataset['Yr.Sold'], month=dataset['Mo.Sold'], day=1)).apply(lambda x: x.timestamp())

### Train-Test Split

- A partir de agora, usaremos apenas o dataset de treino, a partição de teste será tratada como se não existisse ainda.
- O dataset total será dividido em uma proporção 80/20, uma vez que temos poucos dados (2930 no total).
- Por não se tratar de uma série temporal, podemos aplicar uma aleatoriedade na partição.

In [8]:
X, y = dataset.drop('SalePrice', axis=1), dataset.loc[:, 'SalePrice']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=SEED)

## Etapa 1

### Análise Exploratória

Nesta parte, iremos fazer uma análise global dos dados, apenas para garantir a integridade destes.  
Assim sendo, iremos procurar entender quais são as features e target, quais são seus respectivos tipos e buscar outras informações como:
- Dados nulos
- Dados duplicados
- Outliers
- Spikes
- Erros grosseiros

Além disso, iremos buscar saber a distribuição e a "cara" de cada variável.

In [10]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 85 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   MS.SubClass       2930 non-null   int64  
 1   MS.Zoning         2930 non-null   object 
 2   Lot.Frontage      2440 non-null   float64
 3   Lot.Area          2930 non-null   int64  
 4   Street            2930 non-null   object 
 5   Alley             198 non-null    object 
 6   Lot.Shape         2930 non-null   object 
 7   Land.Contour      2930 non-null   object 
 8   Utilities         2930 non-null   object 
 9   Lot.Config        2930 non-null   object 
 10  Land.Slope        2930 non-null   object 
 11  Neighborhood      2930 non-null   object 
 12  Condition.1       2930 non-null   object 
 13  Condition.2       2930 non-null   object 
 14  Bldg.Type         2930 non-null   object 
 15  House.Style       2930 non-null   object 
 16  Overall.Qual      2930 non-null   int64  


In [11]:
dataset.describe(include='all')

MS.SubClass MS.Zoning  Lot.Frontage       Lot.Area Street Alley  \
count   2930.000000      2930   2440.000000    2930.000000   2930   198   
unique          NaN         7           NaN            NaN      2     2   
top             NaN        RL           NaN            NaN   Pave  Grvl   
freq            NaN      2273           NaN            NaN   2918   120   
mean      57.387372       NaN     69.224590   10147.921843    NaN   NaN   
std       42.638025       NaN     23.365335    7880.017759    NaN   NaN   
min       20.000000       NaN     21.000000    1300.000000    NaN   NaN   
25%       20.000000       NaN     58.000000    7440.250000    NaN   NaN   
50%       50.000000       NaN     68.000000    9436.500000    NaN   NaN   
75%       70.000000       NaN     80.000000   11555.250000    NaN   NaN   
max      190.000000       NaN    313.000000  215245.000000    NaN   NaN   

       Lot.Shape Land.Contour Utilities Lot.Config  ...      Mo.Sold  \
count       2930         2930      2930       2930  ...  2930.000000   
unique         4            4         3          5  ...          NaN   
top          Reg          Lvl    AllPub     Inside  ...          NaN   
freq        1859         2633      2927       2140  ...          NaN   
mean         NaN          NaN       NaN        NaN  ...     6.216041   
std          NaN          NaN       NaN        NaN  ...     2.714492   
min          NaN          NaN       NaN        NaN  ...     1.000000   
25%          NaN          NaN       NaN        NaN  ...     4.000000   
50%          NaN          NaN       NaN        NaN  ...     6.000000   
75%          NaN          NaN       NaN        NaN  ...     8.000000   
max          NaN          NaN       NaN        NaN  ...    12.000000   

            Yr.Sold Sale.Type Sale.Condition      SalePrice   Tot.Lot.Area  \
count   2930.000000      2930           2930    2930.000000    2440.000000   
unique          NaN        10              6            NaN            NaN   
top             NaN       WD          Normal            NaN            NaN   
freq            NaN      2536           2413            NaN            NaN   
mean    2007.790444       NaN            NaN  180796.060068    9778.747541   
std        1.316613       NaN            NaN   79886.692357    6435.636376   
min     2006.000000       NaN            NaN   12789.000000    1324.000000   
25%     2007.000000       NaN            NaN  129500.000000    7283.750000   
50%     2008.000000       NaN            NaN  160000.000000    9331.500000   
75%     2009.000000       NaN            NaN  213500.000000   11287.250000   
max     2010.000000       NaN            NaN  755000.000000  215395.000000   

        Bsmt.Tot.Bath  Garage.Area/Cars  Tot.Porch.SF     Date.Sold  
count     2928.000000       2772.000000   2930.000000  2.930000e+03  
unique            NaN               NaN           NaN           NaN  
top               NaN               NaN           NaN           NaN  
freq              NaN               NaN           NaN           NaN  
mean         0.461919        272.233574     89.139590  1.206200e+09  
std          0.520835         61.659127    107.734138  4.105835e+07  
min          0.000000        100.000000      0.000000  1.136074e+09  
25%          0.000000        235.500000      0.000000  1.172707e+09  
50%          0.000000        264.000000     50.000000  1.207008e+09  
75%          1.000000        296.000000    136.000000  1.243814e+09  
max          3.000000        924.000000   1207.000000  1.277942e+09  

[11 rows x 85 columns]

In [12]:
plot_Xtrain(X_train, 'x_train_original.png')

Gráfico salvo em ./graphs/x_train_original.png


In [13]:
# Agora, basta remover os outliers encontrados no target do dataset
X_train, y_train = remove_outliers(X_train, y_train)

In [14]:
cauda_direita = ['Lot.Frontage', 'Lot.Area', 'Mas.Vnr.Area', 'BsmtFin.SF.1', 'BsmtFin.SF.2',
                 'Bsmt.Unf.SF', 'Total.Bsmt.SF', 'X1st.Flr.SF', 'X2nd.Flr.SF', 'Gr.Liv.Area', 
                 'Garage.Area', 'Wood.Deck.SF', 'Open.Porch.SF', 'Enclosed.Porch', 
                 'Screen.Porch', 'X3Ssn.Porch', 'Tot.Lot.Area', 'Garage.Area/Cars', 'Tot.Porch.SF']

categorical = X.select_dtypes(include='object').columns.tolist()
numerical = X.select_dtypes(include='number').columns.tolist()

In [15]:
X_train_log = X_train.copy()
X_train_log[cauda_direita] = np.log1p(X_train_log[cauda_direita])

plot_Xtrain(X_train[numerical], 'x_train_log.png')


Gráfico salvo em ./graphs/x_train_log.png


In [16]:
num_pipe = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
])

c_log_pipe = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('log', FunctionTransformer(np.log1p, validate=False, feature_names_out='one-to-one')),
])

cat_pipe = Pipeline(steps=[
    ('encoder', OneHotEncoder(drop='first', handle_unknown='ignore'))
])

preprocessing_pipeline = ColumnTransformer(
    transformers = [
        ('num', num_pipe, [col for col in numerical if col not in cauda_direita]),
        ('clog', c_log_pipe, cauda_direita),
        ('cat', cat_pipe, categorical)
    ],
    remainder='passthrough'
)

preprocessing_pipeline

ColumnTransformer(remainder='passthrough',
                  transformers=[('num',
                                 Pipeline(steps=[('imputer', SimpleImputer())]),
                                 ['MS.SubClass', 'Overall.Qual', 'Overall.Cond',
                                  'Year.Built', 'Year.Remod.Add',
                                  'Low.Qual.Fin.SF', 'Bsmt.Full.Bath',
                                  'Bsmt.Half.Bath', 'Full.Bath', 'Half.Bath',
                                  'Bedroom.AbvGr', 'Kitchen.AbvGr',
                                  'TotRms.AbvGrd', 'Fireplaces',
                                  'Garage.Yr.Blt', 'Garage.Cars'...
                                  'Land.Contour', 'Utilities', 'Lot.Config',
                                  'Land.Slope', 'Neighborhood', 'Condition.1',
                                  'Condition.2', 'Bldg.Type', 'House.Style',
                                  'Roof.Style', 'Roof.Matl', 'Exterior.1st',
                                  'Exterior.2nd', 'Mas.Vnr.Type', 'Exter.Qual',
                                  'Exter.Cond', 'Foundation', 'Bsmt.Qual',
                                  'Bsmt.Cond', 'Bsmt.Exposure',
                                  'BsmtFin.Type.1', 'BsmtFin.Type.2', 'Heating',
                                  'Heating.QC', 'Central.Air', 'Electrical', ...])])

In [17]:
X_train_transformed = preprocessing_pipeline.fit_transform(X_train)

In [18]:
rnd_clf = RandomForestRegressor(
    n_estimators=700,
    max_leaf_nodes=16,
    random_state=SEED,
    n_jobs=1
)

In [19]:
rnd_clf.fit(X_train_transformed, y_train)

RandomForestRegressor(max_leaf_nodes=16, n_estimators=700, n_jobs=1,
                      random_state=420)

In [20]:
importances = rnd_clf.feature_importances_

feature_names = preprocessing_pipeline.get_feature_names_out()
feature_importances_df = pd.DataFrame(zip(feature_names, importances), columns=['Feature', 'Importance']) \
    .sort_values(by='Importance', ascending=False)

feature_importances_df = aggregate_categorical_importances(feature_importances_df.set_index('Feature'))


X_feats = [feat.split('__')[1] for feat in feature_importances_df[:15].index]

feature_importances_df[:15]

Aggregated_Importance
num__Overall.Qual                 0.695280
clog__Gr.Liv.Area                 0.130847
clog__Total.Bsmt.SF               0.033933
num__Garage.Cars                  0.033929
clog__Garage.Area                 0.028159
clog__X1st.Flr.SF                 0.014843
clog__BsmtFin.SF.1                0.011778
num__Year.Built                   0.011039
clog__Lot.Area                    0.006394
num__Year.Remod.Add               0.004653
cat__Garage.Type                  0.004419
cat__Bsmt.Qual                    0.001909
cat__Kitchen.Qual                 0.001901
clog__X2nd.Flr.SF                 0.001746
cat__MS.Zoning                    0.001711

In [21]:
X_train = X_train[X_feats]

## Parte 3/4/5/6/...

1. Escolher modelos (métodos de stacking inclusos)
    - DummyRegressor
    - LinearRegression
    - Outros modelos básicos
        - Polynomial Features
        - Scalers
    - Pipelines avançadas
        - Utilizar KMeans como fonte de novas features na pipeline
        - Métodos de Ensemble
        
2. Montar GridSearchCV com hiperparâmetros

In [22]:
X_cat = X_train.select_dtypes(include='object').columns.tolist()
X_num = X_train.select_dtypes(include='number').columns.tolist()
X_log = [feat for feat in X_num if feat in cauda_direita]

In [23]:
log_transformer = FunctionTransformer(np.log1p, validate=False, feature_names_out='one-to-one')

In [24]:
cat_pipe = Pipeline(steps=[
    ('encoder', OneHotEncoder(drop='first', handle_unknown='ignore')),
    ('imputer', SimpleImputer(strategy='most_frequent'))
])

num_pipe = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
])

log_pipe = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('log', log_transformer)
])

preprocessing_pipeline = ColumnTransformer(
    transformers = [
        ('num', num_pipe, X_num),
        # ('log', log_pipe, X_log ),
        ('cat', cat_pipe, X_cat)
    ],
    remainder='passthrough'
)

preprocessing_pipeline

ColumnTransformer(remainder='passthrough',
                  transformers=[('num',
                                 Pipeline(steps=[('imputer', SimpleImputer())]),
                                 ['Overall.Qual', 'Gr.Liv.Area',
                                  'Total.Bsmt.SF', 'Garage.Cars', 'Garage.Area',
                                  'X1st.Flr.SF', 'BsmtFin.SF.1', 'Year.Built',
                                  'Lot.Area', 'Year.Remod.Add',
                                  'X2nd.Flr.SF']),
                                ('cat',
                                 Pipeline(steps=[('encoder',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore')),
                                                 ('imputer',
                                                  SimpleImputer(strategy='most_frequent'))]),
                                 ['Garage.Type', 'Bsmt.Qual', 'Kitchen.Qual',
                                  'MS.Zoning'])])

In [25]:
def preprocess(reg, scaling=False,poly=False):
    return Pipeline(steps=[
        ('preprocessing', preprocessing_pipeline),
        ('scaler', StandardScaler() if scaling else 'passthrough'),
        ('poly', PolynomialFeatures(degree=2, include_bias=False) if poly else 'passthrough'),
        ('regressor', reg)
    ])

In [ ]:
param_grid = [{
    'regressor' : [
        preprocess(LinearRegression(), scaling=True, poly=True), 
        DummyRegressor()
    ],

}, {
    'regressor': [
        preprocess(Lasso(), scaling=True, poly=True),
        preprocess(Ridge(), scaling=True, poly=True)
    ],

    'regressor__regressor__alpha': [0.1, 1, 10, 100],
}, {
    'regressor': [
        preprocess(ElasticNet(), scaling=True, poly=True)
    ],

    'regressor__regressor__alpha': [0.1, 1, 10, 100],
    'regressor__regressor__l1_ratio': [0.1, 0.5, 0.9]
}, {
    'regressor': [
        preprocess(RandomForestRegressor())
    ],

    'regressor__regressor__n_estimators': [10, 50, 100],
    'regressor__regressor__max_depth': [None, 10, 20, 30],
    'regressor__regressor__min_samples_split': [2, 5, 10]
}, {
    'regressor': [
        preprocess(GradientBoostingRegressor(), scaling=True, poly=True)
    ],

    'regressor__regressor__n_estimators': [10, 50, 100],
    'regressor__regressor__learning_rate': [0.01, 0.1, 0.2],
    'regressor__regressor__max_depth': [3, 5, 7]
}]

model = Pipeline(steps=[
    ('regressor', LinearRegression())
])

gd_cv = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    scoring='neg_root_mean_squared_error',
    cv=5,
    return_train_score=True,
    n_jobs=-1,
    verbose=1
)

gd_cv

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('regressor', LinearRegression())]),
             n_jobs=-1,
             param_grid=[{'regressor': [Pipeline(steps=[('preprocessing',
                                                         ColumnTransformer(remainder='passthrough',
                                                                           transformers=[('num',
                                                                                          Pipeline(steps=[('imputer',
                                                                                                           SimpleImputer())]),
                                                                                          ['Overall.Qual',
                                                                                           'Gr.Liv.Area',
                                                                                           'Total.Bsmt.SF',
                                                                                           'Garage.Cars',
                                                                                           'Garage.Area',
                                                                                           'X1st.Flr.SF',
                                                                                           'Bsmt...
                                                                                           'MS.Zoning'])])),
                                                        ('scaler',
                                                         StandardScaler()),
                                                        ('poly',
                                                         PolynomialFeatures(include_bias=False)),
                                                        ('regressor',
                                                         GradientBoostingRegressor())])],
                          'regressor__regressor__learning_rate': [0.01, 0.1,
                                                                  0.2],
                          'regressor__regressor__max_depth': [3, 5, 7],
                          'regressor__regressor__n_estimators': [10, 50, 100]}],
             return_train_score=True, scoring='log_mean_squared_error',
             verbose=1)

In [27]:
gd_cv.fit(X_train, y_train)
best_model = gd_cv.best_estimator_


InvalidParameterError: The 'scoring' parameter of GridSearchCV must be a str among {'homogeneity_score', 'neg_max_error', 'neg_mean_poisson_deviance', 'neg_brier_score', 'f1_samples', 'neg_mean_gamma_deviance', 'jaccard_samples', 'adjusted_rand_score', 'recall_macro', 'r2', 'roc_auc_ovo', 'f1_micro', 'neg_root_mean_squared_error', 'recall_micro', 'mutual_info_score', 'accuracy', 'precision_macro', 'jaccard_micro', 'v_measure_score', 'rand_score', 'explained_variance', 'f1', 'precision', 'jaccard_weighted', 'roc_auc_ovr', 'neg_mean_absolute_error', 'neg_mean_squared_error', 'fowlkes_mallows_score', 'recall', 'neg_root_mean_squared_log_error', 'jaccard_macro', 'precision_weighted', 'balanced_accuracy', 'precision_samples', 'matthews_corrcoef', 'roc_auc_ovo_weighted', 'adjusted_mutual_info_score', 'recall_samples', 'f1_macro', 'neg_mean_absolute_percentage_error', 'jaccard', 'positive_likelihood_ratio', 'roc_auc', 'roc_auc_ovr_weighted', 'neg_median_absolute_error', 'top_k_accuracy', 'neg_log_loss', 'average_precision', 'precision_micro', 'normalized_mutual_info_score', 'completeness_score', 'recall_weighted', 'neg_mean_squared_log_error', 'd2_absolute_error_score', 'neg_negative_likelihood_ratio', 'f1_weighted'}, a callable, an instance of 'list', an instance of 'tuple', an instance of 'dict' or None. Got 'log_mean_squared_error' instead.

## Parte 5

In [161]:
import pandas as pd

df = pd.DataFrame(gd_cv.cv_results_) \
    .sort_values(by='rank_test_score') \
    .loc[:, ['params', 'mean_train_score', 'mean_test_score', 'std_test_score']] \
    .assign(
        mean_test_score=lambda df: df['mean_test_score'].apply(lambda x: -x),
        mean_train_score=lambda df: df['mean_train_score'].apply(lambda x: -x),
    )


In [162]:
df

params  mean_train_score  \
72  {'regressor': (ColumnTransformer(remainder='pa...      5.725854e+07   
71  {'regressor': (ColumnTransformer(remainder='pa...      1.012294e+08   
69  {'regressor': (ColumnTransformer(remainder='pa...      1.863647e+08   
41  {'regressor': (ColumnTransformer(remainder='pa...      6.843972e+07   
27  {'regressor': (ColumnTransformer(remainder='pa...      8.895388e+07   
54  {'regressor': (ColumnTransformer(remainder='pa...      8.757652e+07   
51  {'regressor': (ColumnTransformer(remainder='pa...      6.414979e+07   
45  {'regressor': (ColumnTransformer(remainder='pa...      8.698389e+07   
42  {'regressor': (ColumnTransformer(remainder='pa...      6.450168e+07   
24  {'regressor': (ColumnTransformer(remainder='pa...      6.388203e+07   
78  {'regressor': (ColumnTransformer(remainder='pa...      1.235767e+08   
80  {'regressor': (ColumnTransformer(remainder='pa...      5.913741e+07   
77  {'regressor': (ColumnTransformer(remainder='pa...      1.905940e+08   
23  {'regressor': (ColumnTransformer(remainder='pa...      6.769709e+07   
30  {'regressor': (ColumnTransformer(remainder='pa...      1.367329e+08   
56  {'regressor': (ColumnTransformer(remainder='pa...      1.410565e+08   
26  {'regressor': (ColumnTransformer(remainder='pa...      9.103273e+07   
50  {'regressor': (ColumnTransformer(remainder='pa...      6.701322e+07   
68  {'regressor': (ColumnTransformer(remainder='pa...      2.624458e+08   
33  {'regressor': (ColumnTransformer(remainder='pa...      1.052485e+08   
57  {'regressor': (ColumnTransformer(remainder='pa...      1.375903e+08   
53  {'regressor': (ColumnTransformer(remainder='pa...      9.109113e+07   
48  {'regressor': (ColumnTransformer(remainder='pa...      1.377176e+08   
36  {'regressor': (ColumnTransformer(remainder='pa...      1.225073e+08   
44  {'regressor': (ColumnTransformer(remainder='pa...      9.069361e+07   
35  {'regressor': (ColumnTransformer(remainder='pa...      1.250260e+08   
81  {'regressor': (ColumnTransformer(remainder='pa...      2.069504e+07   
39  {'regressor': (ColumnTransformer(remainder='pa...      1.623349e+08   
47  {'regressor': (ColumnTransformer(remainder='pa...      1.435280e+08   
32  {'regressor': (ColumnTransformer(remainder='pa...      1.080618e+08   
29  {'regressor': (ColumnTransformer(remainder='pa...      1.400051e+08   
38  {'regressor': (ColumnTransformer(remainder='pa...      1.645242e+08   
74  {'regressor': (ColumnTransformer(remainder='pa...      1.876645e+07   
75  {'regressor': (ColumnTransformer(remainder='pa...      6.095734e+06   
84  {'regressor': (ColumnTransformer(remainder='pa...      7.862266e+05   
83  {'regressor': (ColumnTransformer(remainder='pa...      6.183596e+06   
49  {'regressor': (ColumnTransformer(remainder='pa...      9.171654e+07   
22  {'regressor': (ColumnTransformer(remainder='pa...      9.188600e+07   
46  {'regressor': (ColumnTransformer(remainder='pa...      1.644106e+08   
25  {'regressor': (ColumnTransformer(remainder='pa...      1.160792e+08   
52  {'regressor': (ColumnTransformer(remainder='pa...      1.143339e+08   
37  {'regressor': (ColumnTransformer(remainder='pa...      1.824925e+08   
28  {'regressor': (ColumnTransformer(remainder='pa...      1.603124e+08   
34  {'regressor': (ColumnTransformer(remainder='pa...      1.480284e+08   
40  {'regressor': (ColumnTransformer(remainder='pa...      9.064282e+07   
43  {'regressor': (ColumnTransformer(remainder='pa...      1.142421e+08   
55  {'regressor': (ColumnTransformer(remainder='pa...      1.632365e+08   
31  {'regressor': (ColumnTransformer(remainder='pa...      1.311754e+08   
79  {'regressor': (ColumnTransformer(remainder='pa...      2.680581e+08   
82  {'regressor': (ColumnTransformer(remainder='pa...      1.219261e+08   
5   {'regressor': (ColumnTransformer(remainder='pa...      3.084136e+08   
12  {'regressor': (ColumnTransformer(remainder='pa...      2.918190e+08   
76  {'regressor': (ColumnTransformer(remainder='pa...      5.124079e+08 

Seleção de modelos com GridSearchCV

In [20]:
param_grid = [{
    'regressor' : [LinearRegression(), DummyRegressor()],
}, {
    'regressor': [Lasso(), Ridge()],
    'alpha': [0.1, 1, 10, 100],
}, {
    'regressor': [ElasticNet()],
    'alpha': [0.1, 1, 10, 100],
    'l1_ratio': [0.1, 0.5, 0.9]
}, {
    'regressor': [RandomForestRegressor()],
    'n_estimators': [10, 50, 100],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}, {
    'regressor': [GradientBoostingRegressor()],
    'n_estimators': [10, 50, 100],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7]
}]